diff --git a/xbmc/interfaces/python/PythonInvoker.cpp b/xbmc/interfaces/python/PythonInvoker.cpp
index 2df2d54da1825..bad86f3621856 100644
--- a/xbmc/interfaces/python/PythonInvoker.cpp
+++ b/xbmc/interfaces/python/PythonInvoker.cpp
@@ -43,6 +43,8 @@
 #include <osdefs.h>
 // clang-format on
 
+#include "XBPython.h"
+
 #include <cassert>
 #include <iterator>
 
@@ -154,21 +156,26 @@ bool CPythonInvoker::execute(const std::string& script, std::vector<std::wstring
   {
     if (!m_threadState)
     {
-#if PY_VERSION_HEX < 0x03070000
-      // this is a TOTAL hack. We need the GIL but we need to borrow a PyThreadState in order to get it
-      // as of Python 3.2 since PyEval_AcquireLock is deprecated
-      extern PyThreadState* savestate;
-      PyEval_RestoreThread(savestate);
-#else
-      PyThreadState* ts = PyInterpreterState_ThreadHead(PyInterpreterState_Main());
-      PyEval_RestoreThread(ts);
-#endif
+      m_mainThreadState =
+          PyThreadState_New(CServiceBroker::GetXBPython().GetMainThreadState()->interp);
+
+      if (m_mainThreadState == nullptr)
+      {
+        CLog::LogF(LOGERROR, "({}, {}) PyThreadState_New failed", GetId(), m_sourceFile);
+        return false;
+      }
+
+      PyThreadState_Swap(m_mainThreadState);
       l_threadState = Py_NewInterpreter();
       PyEval_ReleaseThread(l_threadState);
       if (l_threadState == NULL)
       {
         CLog::Log(LOGERROR, "CPythonInvoker({}, {}): FAILED to get thread m_threadState!", GetId(),
                   m_sourceFile);
+        PyThreadState_Swap(m_mainThreadState);
+        PyThreadState_Clear(m_mainThreadState);
+        PyThreadState_DeleteCurrent();
+        m_mainThreadState = nullptr;
         return false;
       }
       newInterp = true;
@@ -462,7 +469,14 @@ bool CPythonInvoker::stop(bool abort)
       setState(InvokerStateStopping);
       lock.unlock();
 
-      PyEval_RestoreThread((PyThreadState*)m_threadState);
+      PyThreadState* ts = PyThreadState_New(m_threadState->interp);
+      if (ts == nullptr)
+      {
+        CLog::LogF(LOGERROR, "({}, {}) PyThreadState_New failed", GetId(), m_sourceFile);
+        return false;
+      }
+
+      PyEval_RestoreThread(ts);
 
       //tell xbmc.Monitor to call onAbortRequested()
       if (m_addon)
@@ -472,7 +486,8 @@ bool CPythonInvoker::stop(bool abort)
         AbortNotification();
       }
 
-      PyEval_ReleaseThread(m_threadState);
+      PyThreadState_Clear(ts);
+      PyThreadState_DeleteCurrent();
     }
     else
       //Release the lock while waiting for threads to finish
@@ -512,10 +527,17 @@ bool CPythonInvoker::stop(bool abort)
     // so we need to recheck for m_threadState == NULL
     if (m_threadState != NULL)
     {
+      PyThreadState* ts = PyThreadState_New(m_threadState->interp);
+      if (ts == nullptr)
+      {
+        CLog::LogF(LOGERROR, "({}, {}) PyThreadState_New failed", GetId(), m_sourceFile);
+        return false;
+      }
+
       {
         // grabbing the PyLock while holding the m_critical is asking for a deadlock
         CSingleExit ex2(m_critical);
-        PyEval_RestoreThread((PyThreadState*)m_threadState);
+        PyEval_RestoreThread(ts);
       }
 
 
@@ -532,7 +554,8 @@ bool CPythonInvoker::stop(bool abort)
       // If a dialog entered its doModal(), we need to wake it to see the exception
       pulseGlobalEvent();
 
-      PyEval_ReleaseThread(m_threadState);
+      PyThreadState_Clear(ts);
+      PyThreadState_DeleteCurrent();
     }
     lock.unlock();
 
@@ -583,12 +606,10 @@ void CPythonInvoker::onExecutionDone()
     // unregister the language hook
     m_languageHook->UnregisterMe();
 
-#if PY_VERSION_HEX < 0x03070000
-    PyEval_ReleaseLock();
-#else
-    PyThreadState_Swap(PyInterpreterState_ThreadHead(PyInterpreterState_Main()));
-    PyEval_SaveThread();
-#endif
+    PyThreadState_Swap(m_mainThreadState);
+    PyThreadState_Clear(m_mainThreadState);
+    PyThreadState_DeleteCurrent();
+    m_mainThreadState = nullptr;
 
     // set stopped event - this allows ::stop to run and kill remaining threads
     // this event has to be fired without holding m_critical
diff --git a/xbmc/interfaces/python/PythonInvoker.h b/xbmc/interfaces/python/PythonInvoker.h
index 052cc31677fd2..30e79ce14326c 100644
--- a/xbmc/interfaces/python/PythonInvoker.h
+++ b/xbmc/interfaces/python/PythonInvoker.h
@@ -61,6 +61,7 @@ class CPythonInvoker : public ILanguageInvoker
   FILE* PyFile_AsFileWithMode(PyObject* py_file, const char* mode);
 
   PyThreadState* m_threadState;
+  PyThreadState* m_mainThreadState{nullptr};
   bool m_stop = false;
   CEvent m_stoppedEvent;
 
diff --git a/xbmc/interfaces/python/XBPython.cpp b/xbmc/interfaces/python/XBPython.cpp
index 6681447b982c7..4aafb21af0a98 100644
--- a/xbmc/interfaces/python/XBPython.cpp
+++ b/xbmc/interfaces/python/XBPython.cpp
@@ -34,9 +34,6 @@
 
 #include <algorithm>
 
-// Only required for Py3 < 3.7
-PyThreadState* savestate;
-
 bool XBPython::m_bInitialized = false;
 
 XBPython::XBPython()
@@ -49,11 +46,10 @@ XBPython::~XBPython()
   XBMC_TRACE;
   CServiceBroker::GetAnnouncementManager()->RemoveAnnouncer(this);
 
-#if PY_VERSION_HEX >= 0x03070000
   if (Py_IsInitialized())
   {
     // Switch to the main interpreter thread before finalizing
-    PyThreadState_Swap(PyInterpreterState_ThreadHead(PyInterpreterState_Main()));
+    PyThreadState_Swap(m_mainThreadState);
 
     // Clear all loaded modules to prevent circular references
     PyObject* modules = PyImport_GetModuleDict();
@@ -61,7 +57,6 @@ XBPython::~XBPython()
 
     Py_Finalize();
   }
-#endif
 }
 
 #define LOCK_AND_COPY(type, dest, src) \
@@ -512,22 +507,7 @@ bool XBPython::OnScriptInitialized(ILanguageInvoker* invoker)
 #endif
 
     Py_Initialize();
-
-#if PY_VERSION_HEX < 0x03070000
-    // Python >= 3.7 Py_Initialize implicitly calls PyEval_InitThreads
-    // Python < 3.7 we have to manually call initthreads.
-    // PyEval_InitThreads is a no-op on subsequent calls, No need to wrap in
-    // PyEval_ThreadsInitialized() check
-    PyEval_InitThreads();
-#endif
-
-    // Acquire GIL if thread doesn't currently hold.
-    if (!PyGILState_Check())
-      PyEval_RestoreThread((PyThreadState*)m_mainThreadState);
-
-    if (!(m_mainThreadState = PyThreadState_Get()))
-      CLog::Log(LOGERROR, "Python threadstate is NULL.");
-    savestate = PyEval_SaveThread();
+    m_mainThreadState = PyEval_SaveThread();
 
     m_bInitialized = true;
   }
diff --git a/xbmc/interfaces/python/XBPython.h b/xbmc/interfaces/python/XBPython.h
index e54b4a2193433..e14c172463f65 100644
--- a/xbmc/interfaces/python/XBPython.h
+++ b/xbmc/interfaces/python/XBPython.h
@@ -21,6 +21,8 @@
 class CPythonInvoker;
 class CVariant;
 
+typedef struct _ts PyThreadState;
+
 typedef struct
 {
   int id;
@@ -100,13 +102,15 @@ class XBPython : public IPlayerCallback,
   void OnScriptFinalized(ILanguageInvoker* invoker) override;
   ILanguageInvoker* CreateInvoker() override;
 
+  PyThreadState* GetMainThreadState() const { return m_mainThreadState; }
+
   bool WaitForEvent(CEvent& hEvent, unsigned int milliseconds);
 
 private:
   static bool m_bInitialized; // whether global python runtime was already initialized
 
   CCriticalSection m_critSection;
-  void* m_mainThreadState{nullptr};
+  PyThreadState* m_mainThreadState{nullptr};
   int m_iDllScriptCounter{0}; // to keep track of the total scripts running that need the dll
 
   //Vector with list of threads used for running scripts
